topi/python/topi/nn/winograd_util.py

```diff
@@ -0,0 +1,151 @@
+        F = np.fromfunction(np.vectorize(f), (n-1, n-1), dtype=int)
+        f = lambda i, j: -a[i]**(n-1)
+        t = np.fromfunction(np.vectorize(f), (n-1, 1), dtype=int)
+        T = np.append(np.eye(n-1), t, axis=1)
+
+        return np.append(F.T.dot(T), np.array([np.eye(n)[n-1]]), axis=0)
+
+    alpha = n + r - 1
+
+    f = _F_m(a, alpha)
+
+    if f[0, 0] < 0:
+        f[0, :] *= -1
+
+    A = _A_m(a, alpha, n)
+
+    G = _A_m(a, alpha, r).T
+    G = G.dot(np.linalg.inv(f)).T
+
+    B = _B_m(a, alpha)
+    B = B.dot(f.T)
+
+    return (A, B, G)
+
+def _interpolation_points(degree):
+    """Propose filter points"""
+
+    assert 2 < degree < 18
+
+    # Default interpolation lookup table
+    #
+    # [1] Error Analysis and Improving the Accuracy of Winograd Convolution for Deep Neural Networks
+    #     Barbara Barabasz, Andrew Anderson, Kirk M. Soodhalter, David Gregg
+    #     https://arxiv.org/abs/1803.10986
+    #
+
+    # pylint: disable=bad-whitespace,line-too-long
+    in_pts = [
+        #   {invalid}
+        [],
+        #01 {E=4.63E-08 on conv2d  [1]}
+        [],
+        #02 {E=7.65E-08 on F( 2,3) [1]}
+        [0,   -1,    1],
+        #03 {E=2.35E-07 on F( 3,3) [1]}
+        [0,   -1,    1,  1/2],
+        #04 {E=3.29E-07 on F( 4,3) [1]}
+        [0,   -1,    1,  1/2,   -2],
+        #05 {E=6.81E-07 on F( 5,3) [1]}
+        [0,   -1,    1,  1/2,   -2, -1/2],
+        #06 {E=8.79E-07 on F( 6,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2],
+        #07 {E=3.71E-06 on F( 7,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2, -1/4],
+        #08 {E=7.35E-06 on F( 8,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2, -1/4,    4],
+        #09 {E=2.20E-05 on F( 9,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2, -1/4,  3/4, -4/3],
+        #10 {E=3.22E-05 on F(10,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2, -1/4,    4,  3/4, -4/3],
+        #11 {E=1.09E-04 on F(11,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2, -1/4,    4,  3/4, -4/3,  1/4],
+        #12 {E=1.99E-04 on F(12,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2, -1/4,    4,  1/4, -3/4,  4/3,   -4],
+        #13 {E=5.54E-04 on F(13,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2, -1/4,    4,  1/4, -3/4,  4/3,  3/4, -4/3],
+        #14 {E=8.80E-04 on F(14,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2, -1/4,    4,  1/4, -3/4,  4/3,   -4,  3/4, -4/3],
+        #15 {E=1.07E-02 on F(15,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2, -1/4,    4,  1/4, -3/4,  4/3,   -4,  2/3, -3/2,  3/2],
+        #16 {E=1.93E-02 on F(16,3) [1]}
+        [0,   -1,    1,  1/2, -1/2,    2,   -2, -1/4,    4,  1/4, -3/4,  4/3,   -4,  2/3, -3/2, -2/3,  3/2]
+    ] # pylint: enable=bad-whitespace,line-too-long
+
+    return np.array(in_pts[degree-1], dtype=np.float64)
+
+def winograd_transform_matrices(tile_size, kernel_size, out_dtype):
```

Comment:
I'm not that familiar with the whole codebase, so the following comment might be irrelevant.

This function performs non-trivial compute, which should probably be memoized on some level.

Reply:
@zlateski ,
  Not sure how to do the memoization. We have a ```@memoize``` decorator, it is used only in test scripts so far and i see it can memoize only functions with no arguments on them.

Reply:
@zlateski,
 Also, if use ```@memoize``` can't see too much yield , (it's a pickle file at all) i think the overhead of IO would be greater for such small data.

Reply:
@zlateski  Very good point!  Actually, memoize only write to file when the program exists. Follow up on #3687